File: solved/141.LinkedListCycle.cpp
#include <map>
#include <iostream>
#include <algorithm>
using namespace std;

class Solution {
public:
  bool hasCycle(ListNode *head) {
    map<ListNode*, bool> found;

    while (head != nullptr) {
      if (found[head]) {
        return true;
      }

      found[head] = true;

      head = head->next;
    }
    return false;
  }
};

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(nullptr) {}
};

File: solved/141.LinkedListCycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
  def hasCycle(self, head: Optional[ListNode]) -> bool:
    found = set()
    cur = head
    while cur:
      if cur in found:
        return True
      found.add(cur)
      cur = cur.next
    return False
